### ISSUE TITLE
Rule Change: no-shadow hoist option strictless level between "all" and "functions"

### ISSUE THREAD

**Issue by seansfkelley** [eslint/eslint#17233]
### What rule do you want to change?

no-shadow

### What change to do you want to make?

Generate fewer warnings

### How do you think the change should be implemented?

Other

### Example code

```js
function scope() {
  if (...) {
    const a = ...;
    return;
  }

  if (...) {
    const b = ...;
  }

  const a = ...;
  const b = ...;
}
```

### What does the rule currently do for this code?

[When `hoist: "all"`, reports a violation on the first declarations of both `a` and `b`.](https://eslint.org/play/#eyJ0ZXh0IjoiZnVuY3Rpb24gc2NvcGUoKSB7XG4gIGlmIChmb28pIHtcbiAgICBjb25zdCBhID0gZm9vO1xuICAgIHJldHVybjtcbiAgfVxuXG4gIGlmIChmb28pIHtcbiAgICBjb25zdCBiID0gZm9vO1xuICB9XG5cbiAgY29uc3QgYSA9IGZvbztcbiAgY29uc3QgYiA9IGZvbztcbn0iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOiJsYXRlc3QiLCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7Im5vLXNoYWRvdyI6WyJlcnJvciIseyJob2lzdCI6ImFsbCJ9XX0sImVudiI6eyJlczYiOnRydWV9fX0=)

[When `hoist: "functions"`, no violations are reported.](https://eslint.org/play/#eyJ0ZXh0IjoiZnVuY3Rpb24gc2NvcGUoKSB7XG4gIGlmIChmb28pIHtcbiAgICBjb25zdCBhID0gZm9vO1xuICAgIHJldHVybjtcbiAgfVxuXG4gIGlmIChmb28pIHtcbiAgICBjb25zdCBiID0gZm9vO1xuICB9XG5cbiAgY29uc3QgYSA9IGZvbztcbiAgY29uc3QgYiA9IGZvbztcbn0iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOiJsYXRlc3QiLCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7Im5vLXNoYWRvdyI6WyJlcnJvciIseyJob2lzdCI6ImZ1bmN0aW9ucyJ9XX0sImVudiI6eyJlczYiOnRydWV9fX0=)

### What will the rule do after it's changed?

No change to `hoist: "all"` or `hoist: "functions"`. When set to a new value (placeholder name!) `hoist: "functions-strict"`, a violation is reported _only_ for `b`. This is because the `if` that the first declaration of `a` is inside definitely returns before the next `a` is initialized, so there's no ambiguity about their lifetimes (they do not overlap).

### Participation

- [ ] I am willing to submit a pull request to implement this change.

### Additional comments

There would be no change if the outer `a` were defined before the inner: that is always a violation under all three settings. It would still be a violation if the linter can't determine if the inner block definitely returns before the outer scope is reached again.

**Comment by aladdin-add** [comment]
Thanks for the input. Per our [rules policies](https://eslint.org/blog/2020/05/changes-to-rules-policies), we can only accept new rules if they relate to new ECMAScript features that reached stage 4 in the preceding 12 months, so this proposal doesn't qualify. You can always make a custom rule and publish a plugin.
